protocol/verification.go

```diff
@@ -0,0 +1,104 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+func (res *ErrorResponse) Verify(string, []byte, sign.PublicKey) ErrorCode {
+	return res.Error
+}
+
+func (df *DirectoryProof) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	// step 1. verify TB's signature
+	if df.TB != nil {
+		if !signKey.Verify(df.TB.Serialize(df.STR.Signature), df.TB.Signature) {
+			return ErrorBadSignature
+		}
+	}
+
+	// step 2. verify STR's signature
+	if !signKey.Verify(df.STR.Serialize(), df.STR.Signature) {
+		return ErrorBadSignature
+	}
+
+	// step 3. verify hash chain
+	if df.STR.Epoch > 0 && savedSTR != nil &&
+		!merkletree.VerifyHashChain(df.STR.PreviousSTRHash, savedSTR) {
+		return ErrorBadSTR
+	}
+
+	// step 4. verify VRF Index
+	vrfKey := vrf.PublicKey(df.STR.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), df.AP.LookupIndex, df.AP.VrfProof) {
+		return ErrorBadVRFProof
+	}
+
+	// step 5. verify auth path
```

Comment:
There should be a verification of the commitment (if it's not null) before you verify the auth path, no?